tock/tock/static/js/components/timecard.js

```diff
@@ -1,336 +1,391 @@
+  let newEntry = entries[entries.length - 1].cloneNode(true);
+
+  // remove display: none style
+  newEntry.querySelector('select').style.display = '';
+
+  let fields = newEntry.querySelectorAll('input, select, textarea');
+  // remove values
+  for (f of fields) {
+    f.value = '';
+  }
 
-    // Wait a tick so the DOM can be updated, in case a new
-    // line item entry had to be created
-    setTimeout(function () {
-      // Set the project and trigger a GUI update
-      $('div.entry:last .entry-project select').val(line.project);
+  newEntry.querySelector('.autocomplete__wrapper').remove();
 
-      // Set the hours and trigger a data update
-      $('div.entry:last .entry-amount input').val(line.hours);
-      $('div.entry:last .entry-amount input').change();
+  const previousNumber = parseInt(newEntry.getAttribute('id').match(/\d/)[0]);
+  const nextNumber = entries.length;
 
-      // Go again with the remaining tock lines
-      addTockLines(tockLines);
-    }, 20);
+  if (nextNumber % 2 == 0) {
+    newEntry.classList.add('even');
+    newEntry.classList.remove('odd');
   } else {
-    // If we're finished processing the list of tock lines,
-    // trigger a change event to update the hours total and
-    // re-sync any local storage.
-    $('div.entry:last select').change();
+    newEntry.classList.add('odd');
+    newEntry.classList.remove('even');
   }
-}
 
-// When you change the hours, redo the totals
-$('body').on('keyup', '.entry-amount input', function () {
-  populateHourTotals();
-});
+  newEntry.setAttribute('id', 'entry-' + nextNumber);
+
+  newEntry
+    .querySelectorAll('input, select, textarea, label')
+    .forEach(function (formItem, i, arr) {
+      if (formItem.tagName.toLowerCase() !== 'label') {
+        let formerID = formItem.getAttribute('id');
+        // we need the input not to have the `-select` on the id -- it interferes with the accessible
+        // autocomplete library
+        let nextID = formerID
+          .replace(previousNumber, nextNumber)
+          .replace('-select', '');
+        formItem.setAttribute('id', nextID);
+
+        let formerName = formItem.getAttribute('name');
+        let nextName = formerName.replace(previousNumber, nextNumber);
+        formItem.setAttribute('name', nextName);
+      } else {
+        let formerFor = formItem.getAttribute('for');
+        let nextFor = formerFor.replace(previousNumber, nextNumber);
+        formItem.setAttribute('for', nextFor);
+      }
+    });
 
-$('body').on('click', '.entry-amount input, .entry-delete input', function () {
-  populateHourTotals();
-});
+  document.querySelector('.entries').appendChild(newEntry);
+
+  accessibleAutocomplete.enhanceSelectElement({
+    showAllValues: true,
+    defaultValue: '',
+    selectElement: newEntry.querySelector('select'),
+    onConfirm: handleConfirm,
+  });
+
+  // Increment the TOTAL_FORMS
+  // FIXME
+  // Why is this needed? the post breaks if it's removed.
```

Comment:
Good question! It's a component of Django's Formsets which we're using to process the inbound N "Django form for a timecard entry". Specifically the management form. https://docs.djangoproject.com/en/2.2/topics/forms/formsets/#understanding-the-managementform

If we _don't_ update this when adding adding project lines, the response behavior get's a bit wonky when there are validation failures. 

I.e. `id_timecardobjects-TOTAL_FORMS = 2`. We submit a timecard with 3 lines that fails validation.  Only the first two project lines will be rendered in the response, the third will be lost.

Reply:
ah, thank you for the explanation! i will update this documenting that so no one else has to scratch their head confused.